src/mono/browser/runtime/cancelable-promise.ts

```diff
@@ -32,17 +38,121 @@ export function wrap_as_cancelable<T>(inner: Promise<T>): ControllablePromise<T>
 export function mono_wasm_cancel_promise(task_holder_gc_handle: GCHandle): void {
     const holder = _lookup_js_owned_object(task_holder_gc_handle) as PromiseHolder;
     mono_assert(!!holder, () => `Expected Promise for GCHandle ${task_holder_gc_handle}`);
+    holder.cancel();
+}
 
-    const promise = holder.promise;
-    loaderHelpers.assertIsControllablePromise(promise);
-    const promise_control = loaderHelpers.getPromiseController(promise);
-    if (holder.isResolved) {
-        // FIXME: something needs to free the GCHandle
-        mono_log_warn("Canceling a promise that has already resolved.");
-        return;
-    }
-    mono_assert(!holder.isCanceled, "This promise already canceled.");
-    holder.isCanceled = true;
-    promise_control.reject(new Error("OperationCanceledException"));
+// NOTE: layout has to match PromiseHolderState in JSHostImplementation.Types.cs
+const enum PromiseHolderState {
+    IsResolving = 0,
 }
 
+const promise_holder_symbol = Symbol.for("wasm promise_holder");
+
+export class PromiseHolder extends ManagedObject {
+    public isResolved = false;
+    public isPosted = false;
+    public data: any = null;
+    public reason: any = null;
+    public constructor(public promise: Promise<any>,
+        private gc_handle: GCHandle,
+        private promiseHolderPtr: number, // could be null for GCV_handle 
+        private res_converter?: MarshalerToCs) {
+        super();
+    }
+
+    // returns true if the promise is being canceled by managed code
+    setIsResolving(): boolean {
+        if (!WasmEnableThreads || this.promiseHolderPtr === 0) {
+            return false;
+        }
+        forceThreadMemoryViewRefresh();
+        if (compareExchangeI32(this.promiseHolderPtr + PromiseHolderState.IsResolving, 1, 0) === 0) {
+            return false;
+        }
+        return true;
+    }
+
+    resolve(data: any) {
+        if (this.isResolved) {
+            return;
+        }
+        if (WasmEnableThreads && this.setIsResolving()) {
+            // we know that cancelation is in flight
+            // because we need to keep the GCHandle alive until until the cancelation arrives
+            // we skip the this resolve and let the cancelation to reject the Task
+            // we store the original data and use it later
+            this.data = data;
```

Comment:
it seems to me like there's a risk of `this.data` being assigned twice and the second caller to do it will blow the data away, based on how this is constructed. maybe i don't understand it properly though

Reply:
JS promise could only resolve once